README.txt
OgreAL Perl bindings
====================

This distribution provides Perl bindings for the OgreAL C++ 3D audio library
by Casey Borders. See http://ogreal.sourceforge.net/ or search for
OgreAL on the OGRE 3D wiki http://www.ogre3d.org/wiki/index.php .

OgreAL integrates the OpenAL audio library with OGRE 3D. That lets you
attach a sound to a node in a 3D scene, so that the sound seems to be
coming from a certain place within the scene, which could be interesting
in games for example.

I was going to just bundle this in with the Perl Ogre module, but putting it
in a separate module allows people to download it only if they need it
(Perl Ogre is getting pretty big already, as it is...)


DEPENDENCIES

Required:

  Ogre

In order to use this module, you'll need to install the Ogre Perl module.

  OgreAL
  libopenal

You also have to have recent (svn) versions of these two libraries.
There are installation notes below.


Optional:

  OIS
  Ogre::BetaGUI

OIS provides a way of obtaining user input. If you have another way to do
that (SDL, for example), or your application doesn't accept user input,
then you don't have to install OIS.

Ogre::BetaGUI allows you to make a simple graphical user interface
with a mouse cursor, windows, buttons, text boxes, and labels.
It is required for running examples/doppler.pl but otherwise
it's up to you.


INSTALLATION

Installation is unfortunately probably going to be difficult
unless you're at ease compiling things from svn repositories.
It would be almost miraculous if you managed to install it
under Windows; I look forward to receiving your patches, though.
(OGRE 3D and OgreAL themselves are cross-platform, but I only
use Ubuntu (GNU/Debian/Linux), and I don't know how to build them
as Perl modules on Windows, or Mac for that matter.)

Here are quick notes on how to install OgreAL.
Note that these show how to install locally, not system wide
(basically remove --prefix and do `make install` as root).


# Install OGRE 3D, etc., as you already have for Perl Ogre.
Should be able to get headers/includes with `pkg-config --cflags OGRE`,
`pkg-config --libs OGRE` (at least on Ubuntu).


# Install OIS 1.0 (important: not 0.99+1.0rc1)
Download from sourceforge, then:
$ tar -zxf ois-1.0.tar.gz
$ cd ois/
$ ./bootstrap
$ ./configure --prefix=$HOME/.ois-install
$ make
$ make install
Env vars for OgreAL (unnecessary):
$ export OIS_CFLAGS="-I$HOME/.ois-install/include"
$ export OIS_LIBS="-L$HOME/.ois-install/lib -lOIS"


# Install OpenAL and ALUT from svn
$ svn checkout http://www.openal.org/repos/openal/trunk openal
$ cd openal/OpenAL-Sample/     # different for MacOSX and Windows
$ ./autogen.sh
[bunch of warnings, bleh]
$ ./configure --prefix=$HOME/.openal-install
$ make
$ make install
# OgreAL deprecated ALUT, but make sure it's the recent version just in case
$ cd ../alut/
$ ./autogen.sh
[bunch of warnings, bleh]
$ ./configure --prefix=$HOME/.openal-install      # same directory
$ make
$ make install
Env vars for OgreAL (unnecessary):
$ export OPENAL_CFLAGS="-I$HOME/.openal-install/include"
$ export OPENAL_LIBS="-L$HOME/.openal-install/lib -lopenal -lalut"

Note: on my system (Ubuntu) I had to add a file ~/.openalrc
with the following in it:

(define devices '(alsa dsp arts esd native null))

Your mileage may vary.


# Make sure OgreAL sees the right pkg-config files, just in case
# (this part is necessary if you didn't set the above env vars)
$ export PKG_CONFIG_PATH=$HOME/.ois-install/lib/pkgconfig:$PKG_CONFIG_PATH
$ export PKG_CONFIG_PATH=$HOME/.openal-install/lib/pkgconfig:$PKG_CONFIG_PATH


# Install OgreAL (0.2.5 as of 14 Dec 2007)
$ svn co https://ogreal.svn.sourceforge.net/svnroot/ogreal/trunk/OgreAL-Eihort
$ cd OgreAL-Eihort/
$ ./bootstrap
$ ./configure --prefix=$HOME/.ogreal-install
$ make
$ make install

Try some demos:
$ cd Demos/bin/
$ ./Basic

Note: Linux support is still being worked on. See the section on OgreAL
in the ogre3d wiki.


Now that OgreAL is installed, we can try the Perl module.
If you followed the above instructions, OgreAL installed a .pc file
that can make things easier. Just do this:

$ export PKG_CONFIG_PATH=$HOME/.ogreal-install/lib/pkgconfig:$PKG_CONFIG_PATH

then pkg-config can find the C++ headers and libraries.
Otherwise, set these environment variables:

 OGREAL_CFLAGS
 OGREAL_LIBS

Set them to what? Uhhh...?

Then finally build the Perl module normally:

   perl Makefile.PL
   make
   make test
   make install

Note: you might be asked questions during the install. It is
VERY IMPORTANT that you have either the PKG_CONFIG_PATH
or all of the following environment variables set before building:
OGREAL_CFLAGS, OGREAL_LIBS, OPENAL_CFLAGS, OPENAL_LIBS.
Failure to do that correctly may mean that Ogre::AL gets linked
to older versions, particularly of libopenal; this will cause
frustrating and confusing errors.

You might have to edit Makefile.PL to get it to work for your system.
If so, please let me know.


BUGS

No doubt... Please report any bugs/suggestions to <slanning@cpan.org>.


COPYRIGHT AND LICENCE

Copyright 2007, Scott Lanning. All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


OgreAL is copyright Casey Borders and is currently (14 Dec 2007)
distributed under the LGPL license. See the OgreAL distribution
for more details.


Changes.txt
0.03  2009-10-05 | added multichannel.pl example

0.02  2009-10-04 | update for Ogre 1.6 and OIS 1.2
- requires Ogre 1.6 and OIS 1.2 now,
  no backward-compatibility intended

0.01  2007-12-15 | initial release


examples/basic.pl
#!/usr/bin/perl
# This is OgreAL's "Basic" demo ported to Perl.
# Read README.txt in this directory for how to set
# up the resources for this demo.
#
# Search for "soundManager" below, as the rest is not relevant to Ogre::AL
# (unlike the Ogre samples, this one doesn't use Ogre::ExampleApplication
# or ExampleFrameListener)
# See Demos/Basic/ in the OgreAL distribution.


# Note: as always, I put the packages in one file for the examples here,
# but you could put them in separate files.


package DeviceListener;
# implements FrameListener, WindowEventListener, MouseListener, KeyListener

use strict;
use warnings;

use Ogre 0.35;
use Ogre::Degree;
use Ogre::Vector3;
use Ogre::WindowEventUtilities;

use Ogre::AL;
use Ogre::AL::SoundManager;

use OIS 0.04;
use OIS::InputManager;
use OIS::Keyboard qw(:KeyCode);
use OIS::Mouse qw(:MouseButtonID);

use constant ROTATION_SPEED => 10;
use constant MOVEMENT_SPEED => 500;


sub new {
    my ($pkg, $win, $cam, $sceneMgr) = @_;

    my $self = bless {
        mSceneMgr => $sceneMgr,
        camera => $cam,
        mWindow => $win,
        mPitchNode => $cam->getParentSceneNode,
        mContinue => 1,
        mDirection => Ogre::Vector3->new(0, 0, 0),
        yaw => 0,
        pitch => 0,
        mMouse => undef,
        mKeyboard => undef,
        mInputManager => undef,
        mNumScreenShots => 0,
    }, $pkg;

    $self->{mCamNode} = $self->{mPitchNode}->getParentSceneNode;

    # Initialize OIS (Perl version differs somewhat from C++)
    my $windowHnd = $win->getCustomAttributePtr('WINDOW');
    $self->{mInputManager} = OIS::InputManager->createInputSystemPtr($windowHnd);

    if ($self->{mInputManager}->numMice() > 0) {
        # note again this is a little different than in C++
        $self->{mMouse} = $self->{mInputManager}->createInputObjectMouse(1);
        $self->{mMouse}->setEventCallback($self);
    }

    # some checks just in case you had OIS already installed and it wasn't really version 1.0.0
    my $numkbs = $self->{mInputManager}->can('numKeyboards');
    $numkbs = $self->{mInputManager}->can('numKeyBoards') if not defined $numkbs;
    if ($self->{mInputManager}->$numkbs > 0) {
        # note again this is a little different than in C++
        $self->{mKeyboard} = $self->{mInputManager}->createInputObjectKeyboard(1);
        $self->{mKeyboard}->setEventCallback($self);
    }

    $self->windowResized($win);
    Ogre::WindowEventUtilities->addWindowEventListener($win, $self);

    return $self;
}

# note: until I figure out how to get DESTROY to work properly
# (see Ogre::ExampleFrameListener), whenever you close an app it will
# disable autorepeat. If you have an XWindow system (e.g. Linux),
# you can turn it back on with for example:  xset r rate 300 30
# Thanks to akem on #ogre3d for pointing this out.

# These are the same as in Ogre::ExampleFrameListener
sub windowResized {
    my ($self, $win) = @_;

    my ($width, $height) = $win->getMetrics();
    my $mousestate = $self->{mMouse}->getMouseState();

    # note: in C++ this would be like  mousestate.width = width;
    $mousestate->setWidth($width);
    $mousestate->setHeight($height);
}
sub windowClosed {
    my ($self, $win) = @_;

    # note: NEED TO IMPLEMENT overload == operator (etc...)
    # if ($win == $self->{mWindow}) {
    if ($win->getName == $self->{mWindow}->getName) {
        if ($self->{mInputManager}) {
            my $im = $self->{mInputManager};
            if ($self->{mMouse}) {
                $im->destroyInputObject($self->{mMouse});
                delete $self->{mMouse};
            }
            if ($self->{mKeyboard}) {
                $im->destroyInputObject($self->{mKeyboard});
                delete $self->{mMouse};
            }

            OIS::InputManager->destroyInputSystem($im);
            delete $self->{mInputManager};
        }
    }
}

sub frameStarted {
    my ($self, $evt) = @_;
    my $mouse = $self->{mMouse};
    my $keyboard = $self->{mKeyboard};

    return 0 if $self->{mWindow}->isClosed;

    $keyboard->capture();
    $mouse->capture();

    # rotate camera
    $self->{mCamNode}->yaw($self->{yaw} * $evt->timeSinceLastFrame);
    $self->{mCamNode}->pitch($self->{pitch} * $evt->timeSinceLastFrame);

    $self->{yaw} = 0;
    $self->{pitch} = 0;

    # move camera
    # Quaternion * Vector3 * Real   =  fancy!
    $self->{mCamNode}->translate($self->{mPitchNode}->_getDerivedOrientation * $self->{mDirection} * $evt->timeSinceLastFrame);

    # update stats
    my $om = Ogre::OverlayManager->getSingletonPtr();
    my $taname = $om->getOverlayElement("TextAreaName");
    $taname->setCaption($self->{mWindow}->getAverageFPS
                          . "\n\nRoar       Music:\nF1 = Play   F5 = Play\nF2 = Pause F6 = Pause\nF3 = Stop   F7 = Stop");

    return $self->{mContinue};
}

sub mouseMoved {
    my ($self, $arg) = @_;
    my $ms = $self->{mMouse}->getMouseState();

    $self->{yaw} = - Ogre::Degree->new($ms->X->rel * ROTATION_SPEED);
    $self->{pitch} = - Ogre::Degree->new($ms->Y->rel * ROTATION_SPEED);

    return 1;
}

sub keyPressed {
    my ($self, $arg) = @_;

    # xxx: I still haven't fixed these OIS constants....
    my $key = $arg->key;
    if ($key == OIS::Keyboard->KC_ESCAPE) {
        $self->{mContinue} = 0;
    }
    elsif ($key == OIS::Keyboard->KC_SYSRQ) {
        my $ss = "screenshot_" . ($self->{mNumScreenShots}++) . ".png";
        $self->{mWindow}->writeContentsToFile($ss);
    }
    elsif ($key == OIS::Keyboard->KC_F1) {
        Ogre::AL::SoundManager->getSingletonPtr->getSound("Roar")->play();
    }
    elsif ($key == OIS::Keyboard->KC_F2) {
        Ogre::AL::SoundManager->getSingletonPtr->getSound("Roar")->pause();
    }
    elsif ($key == OIS::Keyboard->KC_F3) {
        Ogre::AL::SoundManager->getSingletonPtr->getSound("Roar")->stop();
    }
    elsif ($key == OIS::Keyboard->KC_F5) {
        Ogre::AL::SoundManager->getSingletonPtr->getSound("ZeroFactor")->play();
    }
    elsif ($key == OIS::Keyboard->KC_F6) {
        Ogre::AL::SoundManager->getSingletonPtr->getSound("ZeroFactor")->pause();
    }
    elsif ($key == OIS::Keyboard->KC_F7) {
        Ogre::AL::SoundManager->getSingletonPtr->getSound("ZeroFactor")->stop();
    }
    elsif ($key == OIS::Keyboard->KC_UP || $key == OIS::Keyboard->KC_W) {
        # a little weird due to crappy Perl bindings
        $self->{mDirection}->setZ($self->{mDirection}->z - MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_DOWN || $key == OIS::Keyboard->KC_S) {
        $self->{mDirection}->setZ($self->{mDirection}->z + MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_LEFT || $key == OIS::Keyboard->KC_A) {
        # a little weird due to crappy Perl bindings
        $self->{mDirection}->setX($self->{mDirection}->x - MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_RIGHT || $key == OIS::Keyboard->KC_D) {
        $self->{mDirection}->setX($self->{mDirection}->x + MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_PGDOWN || $key == OIS::Keyboard->KC_Q) {
        $self->{mDirection}->setY($self->{mDirection}->y - MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_PGUP || $key == OIS::Keyboard->KC_E) {
        # a little weird due to crappy Perl bindings
        $self->{mDirection}->setY($self->{mDirection}->y + MOVEMENT_SPEED);
    }

    return 1;
}

sub keyReleased {
    my ($self, $arg) = @_;

    # xxx: I still haven't fixed these OIS constants....
    my $key = $arg->key;

    if ($key == OIS::Keyboard->KC_UP || $key == OIS::Keyboard->KC_W) {
        # a little weird due to crappy Perl bindings
        $self->{mDirection}->setZ($self->{mDirection}->z + MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_DOWN || $key == OIS::Keyboard->KC_S) {
        $self->{mDirection}->setZ($self->{mDirection}->z - MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_LEFT || $key == OIS::Keyboard->KC_A) {
        # a little weird due to crappy Perl bindings
        $self->{mDirection}->setX($self->{mDirection}->x + MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_RIGHT || $key == OIS::Keyboard->KC_D) {
        $self->{mDirection}->setX($self->{mDirection}->x - MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_PGDOWN || $key == OIS::Keyboard->KC_Q) {
        $self->{mDirection}->setY($self->{mDirection}->y + MOVEMENT_SPEED);
    }
    elsif ($key == OIS::Keyboard->KC_PGUP || $key == OIS::Keyboard->KC_E) {
        # a little weird due to crappy Perl bindings
        $self->{mDirection}->setY($self->{mDirection}->y - MOVEMENT_SPEED);
    }
}


1;


package OgreApp;

use strict;
use warnings;

use Ogre 0.35 qw(:SceneType :ShadowTechnique :GuiMetricsMode);
use Ogre::ColourValue;
use Ogre::Degree;
use Ogre::Light qw(:LightTypes);
use Ogre::OverlayManager;
use Ogre::Plane;
use Ogre::ResourceGroupManager qw(:GroupName);
use Ogre::Root;
use Ogre::Vector3;

use Ogre::AL::Sound;
use Ogre::AL::SoundManager;


sub new {
    my ($pkg) = @_;
    my $self = bless {
        root => Ogre::Root->new(),
        win => undef,
        sceneMgr => undef,
        camera => undef,
        soundManager => undef,
    }, $pkg;

    # note: I've rearranged this a little differently than Demos/Basic/,
    # showing how it's similar to Ogre::ExampleApplication
    $self->setupResources();
    $self->configure();
    $self->chooseSceneManager();
    $self->createCamera();
    $self->createViewports();

    $self->{soundManager} = Ogre::AL::SoundManager->new();

    $self->createScene();

    my $listener = DeviceListener->new($self->{win}, $self->{camera}, $self->{sceneMgr});
    $self->{root}->addFrameListener($listener);

    return $self;
}

sub start {
    my ($self) = @_;
    $self->{root}->startRendering();
}

sub createScene {
    my ($self) = @_;

    my $mgr = $self->{sceneMgr};

    # xxx: I have no idea what voodoo makes this work in C++....
    # This is the API:
    # void setSkyBox(bool enable, const String &materialName, Real distance=5000, bool drawFirst=true, const Quaternion &orientation=Quaternion::IDENTITY, const String &groupName=ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME)
    # $mgr->setSkyBox(1, "Sky", 5, 8, 4000);
    $mgr->setSkyBox(1, "Sky", 4000);
    $mgr->setAmbientLight(Ogre::ColourValue->new(0.5, 0.5, 0.5));
    $mgr->setShadowTechnique(SHADOWTYPE_STENCIL_ADDITIVE);

    my $node = $mgr->getRootSceneNode->createChildSceneNode("OgreHead");
    my $ent = $mgr->createEntity("Ogre", "ogrehead.mesh");

    $node->attachObject($ent);
    $node->setPosition(0, 50, 0);

    my $sound = $self->{soundManager}->createSound("Roar", "roar.wav", 1);
    $node->attachObject($sound);
    $sound->play();

    my $bgSound = $self->{soundManager}->createSound("ZeroFactor", "Zero Factor - Untitled.ogg", 1, 1);
    $bgSound->setGain(0.7);
    $bgSound->setRelativeToListener(1);
    $bgSound->play();

    $node = $mgr->getRootSceneNode->createChildSceneNode("CameraNode");
    $node->setPosition(0, 100, 100);
    my $pitchNode = $node->createChildSceneNode("PitchNode");
    $pitchNode->attachObject($self->{camera});
    $pitchNode->attachObject($self->{soundManager}->getListener());
    $pitchNode->pitch(Ogre::Degree->new(-30));

    # Create a ground plane (just like examples/ninja.pl in Perl Ogre distribution)
    my $plane = Ogre::Plane->new(Ogre::Vector3->new(0, 1, 0), 0);
    my $meshmgr = Ogre::MeshManager->getSingletonPtr();
    $meshmgr->createPlane("ground",
                          DEFAULT_RESOURCE_GROUP_NAME,
                          $plane, 15000, 15000, 20, 20, 1, 1, 5, 5,
                          Ogre::Vector3->new(0, 0, 1));
    $ent = $mgr->createEntity("GroundEntity", "ground");
    $mgr->getRootSceneNode->createChildSceneNode->attachObject($ent);
    $ent->setMaterialName("Ground");
    $ent->setCastShadows(0);

    my $light = $mgr->createLight("sun");
    $light->setType(LT_DIRECTIONAL);
    $light->setDirection(-1,-1,-1);

    # Perl note: see Ogre::BetaGUI::GUI createOverlay for a similar example
    # (note also createOverlayContainer is specific to Perl Ogre)
    my $overlayMgr = Ogre::OverlayManager->getSingletonPtr;
    my $panel = $overlayMgr->createOverlayContainer("Panel", "PanelName");
    $panel->setMetricsMode(GMM_PIXELS);
    $panel->setPosition(10, 10);
    $panel->setDimensions(100, 100);

    my $textArea = $overlayMgr->createTextAreaOverlayElement("TextArea", "TextAreaName");
    $textArea->setMetricsMode(GMM_PIXELS);
    $textArea->setPosition(0, 0);
    $textArea->setDimensions(100, 100);
    $textArea->setCharHeight(16);
    $textArea->setFontName("Arial");
    $textArea->setCaption("Hello, World!");

    my $overlay = $overlayMgr->create("AverageFps");
    $overlay->add2D($panel);
    $panel->addChild($textArea);
    $overlay->show();
}

sub setupResources {
    my ($self) = @_;

    my $cf = Ogre::ConfigFile->new();
    $cf->load("resources.cfg");

    # note: this is a Perlish replacement for iterators used in C++
    my $secs = $cf->getSections();
    my $rgm = Ogre::ResourceGroupManager->getSingletonPtr();

    foreach my $sec (@$secs) {
        my $secName = $sec->{name};

        my $settings = $sec->{settings};
        foreach my $setting (@$settings) {
            my ($typeName, $archName) = @$setting;
            $rgm->addResourceLocation($archName, $typeName, $secName);
        }
    }

    $rgm->initialiseAllResourceGroups();
}

sub configure {
    my ($self) = @_;

    # this shows an alternative to the way Ogre::ExampleApplication does it
    unless ($self->{root}->restoreConfig()) {
        unless ($self->{root}->showConfigDialog()) {
            exit;
        }
    }

    $self->{win} = $self->{root}->initialise(1, "Ogre Framework");
}

sub chooseSceneManager {
    my ($self) = @_;
    $self->{sceneMgr} = $self->{root}->createSceneManager(ST_GENERIC, "MainSceneManager");
}

sub createCamera {
    my ($self) = @_;

    $self->{camera} = $self->{sceneMgr}->createCamera("SimpleCamera");
    $self->{camera}->setNearClipDistance(1.0);
}

sub createViewports {
    my ($self) = @_;

    my $vp = $self->{win}->addViewport($self->{camera});
    $vp->setBackgroundColour(Ogre::ColourValue->new(0,0,0));
    $self->{camera}->setAspectRatio($vp->getActualWidth() / $vp->getActualHeight());
}


1;


package main;

# uncomment this if the packages are in separate files:
# use OgreApp;

OgreApp->new->start();


Makefile.PL
# Note: I'm sure this won't work on Windows.
# First of all, I don't think pkg-config gets installed on Windows,
# but you can try setting the environment variables.
# Then you're going to have to compile it using, what, nmake?
# Let me know how to do it if you get it installed.

use strict;
use warnings;
use Config;
use ExtUtils::MakeMaker;
use File::Spec;

main();
exit 0;

sub main {
    my $fixed_args = fixed_args();
    my $varied_args = varied_args();
    my $gpp_warn_args = gpp_warn_args();

    WriteMakefile(%$fixed_args, %$varied_args, %$gpp_warn_args);
}

sub gpp_warn_args {
    # if you're using gcc >= 4.2 and perl < 5.10, you'll probably
    # get warnings like this during `make` :
    #  OgreAL.c:83: warning: deprecated conversion from string constant to 'char*'
    # The C code generated by `xsubpp` uses literal strings (string constants)
    # as args to functions expecting char*, in particular newXSproto.
    # This disables those warnings.

    if ($Config{PERL_REVISION} == 5 && $Config{PERL_VERSION} < 10) {
        if (my_compiler() eq 'g++') {
            my $str = `g++ -dumpversion`;
            unless ($?) {
                chomp $str;
                my ($v, $s) = split /\./, $str;
                if (($v == 4 && $s >= 2) || $v > 4) {
                    return {'CCFLAGS' => '-Wno-write-strings'};
                }
            }
        }
    }

    # there will be no warnings, or we'll just tolerate them
    return {};
}

sub my_compiler {
    return $ENV{'CXX'} || 'g++';
}

sub fixed_args {
    my ($pkg) = @_;

    return {
        'NAME'          => 'Ogre::AL',
        'VERSION_FROM'  => 'lib/Ogre/AL.pm',
        'AUTHOR'        => 'Scott Lanning <slanning@cpan.org>',
        'LD'            => '$(CC)',
        'OBJECT'        => '$(O_FILES)',
        'MAN3PODS'      => {},
        'XSOPT'         => '-C++',
    };
}

sub varied_args {
    my ($been_there) = @_;
    my (@cflags, @libs);

    # Find out if OgreAL is installed

    # any env vars take priority
    if (ogreal_envvars_set()) {
        push @cflags, $ENV{OGREAL_CFLAGS};
        push @libs, $ENV{OGREAL_LIBS};

        push @cflags, $ENV{OPENAL_CFLAGS};
        push @libs, $ENV{OPENAL_LIBS};
    }
    # otherwise try to find info with pkg-config
    elsif (pkgconfig_found('OgreAL', '0.2.5') && pkgconfig_found('openal', '1.1.0')) {
        push @cflags, pkg_config('OgreAL', 'cflags');
        push @libs, pkg_config('OgreAL', 'libs');

        push @cflags, pkg_config('openal', 'cflags');
        push @libs, pkg_config('openal', 'libs');
    }

    else {
        if (defined $been_there) {
            error("OgreAL and/or libopenal CFLAGS and LIBS not found, so giving up.",
                  "Be sure to read README.txt!");
        }
        else {
            # give them a chance
            where_am_i();
            # recursion, tricky... it's just to rerun things now that
            # we got the env vars
            return varied_args('got the t-shirt');
        }
    }

    my @defines = ();   # (not used for now)

    my %prereqs = (
        'Test::More' => 0,
        'Ogre' => 0.35,
    );


    # Check if optional Perl modules are installed.

    my $OIS_REQ_VERSION = 0.04;
    unless (eval { require OIS && $OIS::VERSION >= $OIS_REQ_VERSION }) {
        my $msg = "\nNote: the Perl module OIS >= $OIS_REQ_VERSION is not installed,\n"
          . "so you won't be able to run some examples\n"
          . "or use Ogre::ExampleFrameListener.\n"
          . "Installing OIS would be a very good idea\n"
          . "unless you have some other way to handle keyboard and mouse input.\n"
          . "It's fine to install it after installing OgreAL.\n\n"
          . "Do you want to install OIS now?";

        my $val = ExtUtils::MakeMaker::prompt($msg, 'n');
        if ($val =~ /^y/i) {
            $prereqs{'OIS'} = $OIS_REQ_VERSION;
            print "\nOIS >= $OIS_REQ_VERSION added to prerequisites.\n";
        }
    }


    return {
        'PREREQ_PM'         => \%prereqs,
        'CC'                => my_compiler(),
        'INC'               => join(' ', @cflags),
        'LIBS'              => join(' ', @libs),
        (@defines ? ('DEFINE' => join(' ', @defines)) : ()),
    };
}

sub ogreal_envvars_set {
    return exists $ENV{OGREAL_CFLAGS} && exists $ENV{OGREAL_LIBS}
      && exists $ENV{OPENAL_CFLAGS} && exists $ENV{OPENAL_LIBS}
}

sub pkgconfig_found {
    my ($pkgname, $version, $fatal) = @_;

    unless (pkg_config_installed()) {
        error("pkg-config not found") if defined $fatal;
    }

    my @errors = check_pkg_config($pkgname, $version);
    error(@errors) if defined $fatal;

    return !@errors;
}

# check out the love, man
sub where_am_i {
    REDO: {
        my $msg = "\nI can't get no.... satisfaction, but I tried... and I tried....\n"
          . "Note: you probably need to read README.txt first.\n\n"
          . "I need some help here. You probably have OgreAL.pc (for pkg-config)\n"
          . "installed somewhere off the beaten path, like\n"
          . "$ENV{HOME}/.ogreal-install/lib/pkgconfig/ or wherever.\n\n"
          . "And that's okay....\n\n"
          . "But can can we put that in PKG_CONFIG_PATH? We can\n"
          . "if you give me the full path here.\n"
          . "If you prefer to set CFLAGS and LIBS explicitly,\n"
          . "just hit return. Where is OgreAL.pc ? ";
        my $val = ExtUtils::MakeMaker::prompt($msg, '');
        $val =~ s/(^\s+|\s+$)//g;   # no mo whitespace

        if ($val =~ m{^/}) {
            my $pcfile = File::Spec->catfile($val, 'OgreAL.pc');
            unless (-f $pcfile) {
                print "\nDidn't find '$pcfile'.\n";
                my $giving_up = ExtUtils::MakeMaker::prompt("Do you give up? ", 'Yeah');
                if ($giving_up =~ /^y/i) {
                    last REDO;
                }
            }

            $ENV{PKG_CONFIG_PATH} = '' unless exists $ENV{PKG_CONFIG_PATH};
            $ENV{PKG_CONFIG_PATH} = "$val:" . $ENV{PKG_CONFIG_PATH};
            my $msg = "\n'$val' added to PKG_CONFIG_PATH - Thanks!\n"
              . "One more to check for... openal.\n"
              . "It might be somewhere local, like\n"
              . "$ENV{HOME}/.openal-install/lib/pkgconfig/ or wherever,\n"
              . "probably called openal.pc. If you think it's already in PKG_CONFIG_PATH,\n"
              . "just hit return. Where is openal.pc ? ";
            my $val = ExtUtils::MakeMaker::prompt($msg, '');
            $val =~ s/(^\s+|\s+$)//g;   # no mo whitespace

            if ($val =~ m{^/}) {
                my $pcfile = File::Spec->catfile($val, 'openal.pc');
                unless (-f $pcfile) {
                    print "\nDidn't find '$pcfile'.\n";
                    my $giving_up = ExtUtils::MakeMaker::prompt("Do you give up? ", 'Yeah');
                    if ($giving_up =~ /^y/i) {
                        last REDO;
                    }
                }
            }
            elsif ($val =~ /\S/) {
                print "\nNo, I need a *full* path, absolute, starting with a SLASH.\n";
                redo REDO;
            }
            else {
                print "\nDone with PKG_CONFIG_PATH.\n";
            }
        }
        elsif ($val =~ /\S/) {
            print "\nNo, I need a *full* path, absolute, starting with a SLASH.\n";
            redo REDO;
        }
        else {
            print "\nOkay, let's forget about PKG_CONFIG_PATH, then.\n";
        }
    }

    unless (exists $ENV{PKG_CONFIG_PATH}) {
        print "\nSince you don't have PKG_CONFIG_PATH set,\n"
          . "I have some harder questions for you...\n";

        unless (exists $ENV{OGREAL_CFLAGS} && exists $ENV{OGREAL_LIBS}) {
            print "\nWhat CFLAGS (header locations, -I paths) should we use for OgreAL?\n";
            my $cflags = ExtUtils::MakeMaker::prompt("OGREAL_CFLAGS = ", '');
            if ($cflags =~ /\S/) {
                $ENV{OGREAL_CFLAGS} = $cflags;
            }

            print "\nWhat LIBS (library locations and libraries, -L and -l) should we use for OgreAL?\n";
            my $libs = ExtUtils::MakeMaker::prompt("OGREAL_LIBS = ", '');
            if ($libs =~ /\S/) {
                $ENV{OGREAL_LIBS} = $libs;
            }
        }

        unless (exists $ENV{OPENAL_CFLAGS} && exists $ENV{OPENAL_LIBS}) {
            print "\nWhat CFLAGS (header locations, -I paths) should we use for libopenal?\n";
            my $cflags = ExtUtils::MakeMaker::prompt("OPENAL_CFLAGS = ", '');
            if ($cflags =~ /\S/) {
                $ENV{OPENAL_CFLAGS} = $cflags;
            }

            print "\nWhat LIBS (library locations and libraries, -L and -l) should we use for libopenal?\n";
            my $libs = ExtUtils::MakeMaker::prompt("OPENAL_LIBS = ", '');
            if ($libs =~ /\S/) {
                $ENV{OPENAL_LIBS} = $libs;
            }
        }
    }

    unless (exists($ENV{PKG_CONFIG_PATH})
              or
            (exists($ENV{OGREAL_CFLAGS}) and exists($ENV{OGREAL_LIBS})
               and exists($ENV{OPENAL_CFLAGS}) and exists($ENV{OPENAL_LIBS}) ))
    {
        error("Giving up - don't know where OgreAL and libopenal are installed.");
    }
}

sub pkg_config_installed {
    my $pcver = `pkg-config --version`;
    return $pcver ne '';
}

sub check_pkg_config {
    my ($pkg, $required_version) = @_;

    my $pc = 'pkg-config';
    my @errors = ();

    # Check that pkg-config knows about the package
    my $pkgexists = `$pc --exists $pkg`;
    if ($?) {
        push @errors, "Package $pkg not found by $pc";
    }
    else {
        # Check that the package is the right version
        my $pkgver = `$pc --atleast-version='$required_version' $pkg`;
        if ($?) {
            push @errors, "Package $pkg is not the right version (at least $required_version)";
        }
    }

    return @errors;
}

sub pkg_config {
    my ($pkg, $option, $pc) = @_;

    $pc = 'pkg-config' unless defined $pc;

    my $str = `$pc --$option $pkg`;
    if ($?) {
        error("$pc --$option $pkg: $str");
    }
    else {
        chomp $str;
        return $str;
    }
}

sub error {
    my (@errors) = @_;
    push @errors, "Be sure to read README.txt!";
    die(map { "$_$/" } @errors);
}
